pkg/volume/csi/csi_plugin.go

```diff
@@ -246,16 +248,84 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
 		}
 	}
 
-	// Initializing csiDrivers map and label management channels
+	// Initializing csiDrivers map, label management channels, and migratedDrivers map
 	csiDrivers = csiDriversStore{driversMap: map[string]csiDriver{}}
-	nim = nodeinfomanager.NewNodeInfoManager(host.GetNodeName(), host)
 
-	// TODO(#70514) Init CSINodeInfo object if the CRD exists and create Driver
-	// objects for migrated drivers.
+	var migratedDrivers = map[string](func() bool){
+		csiPlugins.GCEPDDriverName: func() bool {
+			return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) && utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationGCE)
+		},
+		// TODO(leakingtpan): Add AWS migration feature gates and place them here
+	}
+
+	nim = nodeinfomanager.NewNodeInfoManager(host.GetNodeName(), host, migratedDrivers)
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		// This function prevents Kubelet from posting Ready status until CSINodeInfo
+		// is both installed and initialized
+		initializeCSINodeInfo(host)
+	}
 
 	return nil
 }
 
+func initializeCSINodeInfo(host volume.VolumeHost) {
+	kubeClient := host.GetKubeClient()
+	if kubeClient == nil || len(host.GetPodsDir()) == 0 {
```

Comment:
is `len(host.GetPodsDir()) == 0` the canonical way to check if we are not on the kubelet? that seems rather indirect

Reply:
I don't know about "canonical" since I don't see anyone else doing it. But from what I understand these indirect checks are the "best" way right now, see comment: https://github.com/kubernetes/kubernetes/pull/70909#discussion_r242768000